x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/MvExpandOperator.java

```diff
@@ -46,57 +48,171 @@ public String describe() {
-        int[] duplicateFilter = new int[newPositions];
+    private int[] nextDuplicateExpandingFilter(Block expandingBlock, int size) {
+        int[] duplicateFilter = new int[size];
         int n = 0;
-        for (int p = 0; p < expandingBlock.getPositionCount(); p++) {
-            int count = expandingBlock.getValueCount(p);
+        while (nextPositionToProcess < expandingBlock.getPositionCount()) {
+            int count = expandingBlock.getValueCount(nextPositionToProcess);
             int positions = count == 0 ? 1 : count;
-            Arrays.fill(duplicateFilter, n, n + positions, p);
-            n += positions;
+            int toAdd = Math.min(size - n, positions - nextMvToProcess);
+            Arrays.fill(duplicateFilter, n, n + toAdd, nextPositionToProcess);
+            n += toAdd;
+
+            if (n == size) {
+                if (nextMvToProcess + toAdd == positions) {
+                    nextMvToProcess = 0;
+                    nextPositionToProcess++;
+                    if (nextPositionToProcess == expandingBlock.getPositionCount()) {
+                        nextPositionToProcess = 0;
+                        prevCompleted = true;
+                    }
+                } else {
+                    nextMvToProcess = nextMvToProcess + toAdd;
+                }
+                return duplicateFilter;
+            }
+
+            nextMvToProcess = 0;
+            nextPositionToProcess++;
+        }
+
+        if (nextPositionToProcess == expandingBlock.getPositionCount()) {
+            nextPositionToProcess = 0;
+            nextMvToProcess = 0;
+            prevCompleted = true;
         }
-        return duplicateFilter;
+        return n < size ? Arrays.copyOfRange(duplicateFilter, 0, n) : duplicateFilter;
+    }
+
+    @Override
+    public final boolean needsInput() {
+        return prev == null && finished == false;
     }
 
     @Override
-    protected AbstractPageMappingOperator.Status status(int pagesProcessed) {
+    public final void addInput(Page page) {
+        assert prev == null : "has pending input page";
+        prev = page;
+        prevCompleted = false;
+    }
+
+    @Override
+    public final void finish() {
+        finished = true;
+    }
+
+    @Override
+    public final boolean isFinished() {
+        return finished && prev == null;
+    }
+
+    @Override
+    public final Status status() {
         return new Status(pagesProcessed, noops);
     }
 
+    @Override
+    public void close() {
+        if (prev != null) {
+            Releasables.closeExpectNoException(() -> prev.releaseBlocks());
+        }
+    }
+
     @Override
     public String toString() {
         return "MvExpandOperator[channel=" + channel + "]";
     }
 
-    public static final class Status extends AbstractPageMappingOperator.Status {
+    public static final class Status implements Operator.Status {
+
+        private final int pagesProcessed;
```

Comment:
It'd be nice to have this as `pagesIn` and `pagesOut`. That way you can see what the multiplication factor is too.

Reply:
++